### PR TITLE
Make possible variant SONAMEs and symbol versions

### PR DESCRIPTION
**Pull request by vdukhovni**
This small change in the Unix template and shared library build
scripts makes it possible to build "variant" shared libraries.  A
"variant" shared library has a non-default SONAME, and non default
symbol versions.

This makes it possible to build (say) an OpenSSL 1.1.0 library that
can coexist without conflict in the same process address space as
the system's default OpenSSL library which may be OpenSSL 1.0.2.

This is makes it possible to link applications against a newer
OpenSSL library installed in /opt/openssl/1.1 or similar location,
and not risk conflict with an indirectly loaded OpenSSL runtime
that is required by some other dependency.

This has been fully tested under Linux, and builds successfully on
MacOS/X producing variant DYLD names, but it is not obvious to me
whether this is sufficient (along with Darwins non-flat namespace
for libraries) to avoid symbol conflicts on MacOS/X.

Of course the feature is optional and off by default.

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by kaduk** [comment]
I think I'm confused as to how this helps in the 1.0.2/1.1.0 case, since we regenerated (and reset the versions in) the number files for 1.1.0.  There is of course the case of multiple variants of the same branch, which could also easily occur, of course.

**Comment by vdukhovni** [comment]
> @kaduk I think I'm confused as to how this helps in the 1.0.2/1.1.0 case, since we regenerated (and reset the versions in) the number files for 1.1.0. There is of course the case of multiple variants of the same branch, which could also easily occur, of course.

The symbol versions are indeed already per-release in 1.0.2, but we still separate SONAMEs so that applications find the right shared objects.  Also when using a non-default library in /opt or similar on a system that has no system OpenSSL 1.1.x it is good to know that there'll be no problems should a later upgrade bring in a 1.1.x that may conflict.

**Comment by vdukhovni** [comment]
This also supports systems with OpenSSL 1.0.1 in /usr, where IIRC we did not have upstream symbol versioning.

**Comment by vdukhovni** [comment]
> @levitte I would like to see shlib_variant documented in Configurations/README
That's fair, coming up shortly...

**Comment by vdukhovni** [comment]
Pushed commit to address @levitte comment

File: Configurations/README [hunk 1]
```diff
@@ -101,6 +101,25 @@ In each table entry, the following keys are significant:
...
+                           value of the symlink as the SONAME, a target
+                           definition that sets 'shlib_variant => "-abc"' will
+                           create 'libssl.so -> libssl-abc.so.1.1', again with
+                           an SONAME equal to the value of the symlink.  The
+                           symbol versions associated with the variant library
+                           would then be 'OPENSSL_ABC_<version>' rather than
+                           the default 'OPENSSL_<version>'. The string inserted
+                           into symbol versions is obtained by mapping all
+                           letters in the "variant" identifier to upper case
+                           and all non-alphanumeric characters to '_'.
```
Comment (levitte) [inline Configurations/README]: Same note as in the 1.1.0 PR:

It should be noted that this is available on select platforms only.  We only have symbol versions on Linux and Solaris so far, and the SONAME hack is Unix / MacOS X only.  Windows and VMS be damned in this case.

**Comment by vdukhovni** [comment]
And this is this is the one for master.  Is a second review required?

**Comment by vdukhovni** [comment]
Merged. Thanks.

**Comment by kaduk** [comment]
@vdukhovni Er, merged to where, exactly?  I don't see these commits in either the real repo or github.

**Comment by vdukhovni** [comment]
I merged these in gltlab, and was expecting that to sync into github, but it did not happen...

**Comment by richsalz** [comment]
You have to use the repo's on openssl's git server

**Comment by vdukhovni** [comment]
Ok, I switched from gitlab.openssl.org to git.openssl.org, and pushed, but there's one oddity:

```
Counting objects: 7, done.
Delta compression using up to 8 threads.
Compressing objects: 100% (7/7), done.
Writing objects: 100% (7/7), 2.54 KiB | 2.54 MiB/s, done.
Total 7 (delta 6), reused 0 (delta 0)
remote: Trigg openssl@localhost << checkout  | push openssl.git gitlab github
remote: fatal: ..: '..' is outside repository
To git.openssl.org:openssl.git
   a61c15eb9b..e6f38fb817  so_variants-1.1.0 -> OpenSSL_1_1_0-stable
```
The changes seem to have been accepted despite that `fatal` error. Any idea what that's about?

**Comment by richsalz** [comment]
Debugging noise from @levitte's hook scripts; ignore it.  Maybe now that it's public, he'll silence things :)

**Pull request closed without merging**